Atomize

Jragonmiris opened the pull request:
THIS IS NOT READY FOR PULL, RIGHT NOW THIS IS AN RFC.

I modified the graph interfaces and search package to compile (but not test) with a possible new interface. Part of the problem with atomicity is increasingly weird interface names, but the package "feels" a lot better now.

I removed several algorithms until we know what to do with them. I also nuked everything in concrete, we can retrieve the files from using the correct ref (or, if need be, copy them from master), I just wanted to get to the point where search compiled.

This isn't final.

While I'm at it, I think I'd like to add a Path struct to search which sort of formalizes the outputs a bit. Especially after the nightmare trainwreck that was old FloydWarshall. While the branch name is "atomize" think of this more as "general revision for a possible v1.0 (or at least v0.99)".

btracey commented:
I’m still on temporary hiatus, so sorry for partly off topic, but the go team has been using “DO NOT SUBMIT”  to ask for a review of partially ready code. 


> On Apr 29, 2015, at 10:28 PM, Jeff Juozapaitis <notifications@github.com> wrote:
> 
> THIS IS NOT READY FOR PULL, RIGHT NOW THIS IS AN RFC.
> 
> I modified the graph interfaces and search package to compile (but not test) with a possible new interface. Part of the problem with atomicity is increasingly weird interface names, but the package "feels" a lot better now.
> 
> I removed several algorithms until we know what to do with them. I also nuked everything in concrete, we can retrieve the files from using the correct ref (or, if need be, copy them from master), I just wanted to get to the point where search compiled.
> 
> This isn't final.
> 
> While I'm add it, I think I'd like to add a Path struct to search which sort of formalizes the outputs a bit. Especially after the nightmare trainwreck that was old FloydWarshall.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/gonum/graph/pull/43 <https://github.com/gonum/graph/pull/43>
> Commit Summary
> 
> Initial, incomplete set of interfaces
> Updated package documentation
> Added directed and undirected costs, may need improvement.
> TEMPORARILY removed concrete graphs to work on search
> Added package level documentation for concrete.
> Updated A*, BFS
> Updated dijkstra
> Updated BellmanFord
> Temporarily removed Johnson's algorithm until we get concrete and mutable back up and running
> Updated DFS
> Removed obsolete heuristic functions
> Removed copying functions until we work out new mutability; probably shouldn't be in search anyway
> Updated Tarjan
> Updated IsPath and added some verifying interfaces
> Killed MST algorithms until we have mutable stuff back
> Updated dominators
> Fixed a bug in new Dominators
> Updated PostDominators
> Modified interfaces to require degree where it should be trivial to implement
> Modified InOut a bit more
> Decided on an explicit neighbor method
> Made changes to VertexOrdering
> Forgot to add this earlier, interfaces for search-specific stuff
> Killed graph.Edge reference in interaces
> Removed ugly stuff from internals
> Killed floydwarshall until we figure out how dense graphs fit into the new interface structure
> Removed all errors the compiler didn't catch the first time through
> File Changes
> 
> M concrete/concrete.go <https://github.com/gonum/graph/pull/43/files#diff-0> (7)
> D concrete/dense.go <https://github.com/gonum/graph/pull/43/files#diff-1> (151)
> D concrete/densegraph_test.go <https://github.com/gonum/graph/pull/43/files#diff-2> (193)
> D concrete/mutablegr.go <https://github.com/gonum/graph/pull/43/files#diff-3> (224)
> D concrete/mutablegr_test.go <https://github.com/gonum/graph/pull/43/files#diff-4> (43)
> D concrete/mutdir.go <https://github.com/gonum/graph/pull/43/files#diff-5> (263)
> D concrete/mutdir_test.go <https://github.com/gonum/graph/pull/43/files#diff-6> (43)
> D concrete/tilegraph.go <https://github.com/gonum/graph/pull/43/files#diff-7> (234)
> D concrete/tilegraph_test.go <https://github.com/gonum/graph/pull/43/files#diff-8> (108)
> M doc.go <https://github.com/gonum/graph/pull/43/files#diff-9> (36)
> M graph.go <https://github.com/gonum/graph/pull/43/files#diff-10> (283)
> D search/floydwarshall.go <https://github.com/gonum/graph/pull/43/files#diff-11> (230)
> D search/floydwarshall_test.go <https://github.com/gonum/graph/pull/43/files#diff-12> (182)
> M search/graph_search.go <https://github.com/gonum/graph/pull/43/files#diff-13> (345)
> A search/interfaces.go <https://github.com/gonum/graph/pull/43/files#diff-14> (49)
> M search/internals.go <https://github.com/gonum/graph/pull/43/files#diff-15> (85)
> Patch Links:
> 
> https://github.com/gonum/graph/pull/43.patch <https://github.com/gonum/graph/pull/43.patch>
> https://github.com/gonum/graph/pull/43.diff <https://github.com/gonum/graph/pull/43.diff>
> —
> Reply to this email directly or view it on GitHub <https://github.com/gonum/graph/pull/43>.
>

kortschak commented:
I have a counter proposal. It is less invasive and allows us to explore the ideas around separation of interfaces without major breakage (this can be done on a per function basis).

I have not culled all the things that could be removed because of this set of changes, but there are reasonable savings that can come from this.

It can be viewed [here](https://github.com/gonum/graph/tree/api-explore). It is NOT FOR SUBMISSION.

kortschak commented on graph.go:
```diff
@@ -11,197 +11,162 @@ type Node interface {
-// otherwise.
-type Edge interface {
+// An unordered group of two nodes
+type UndirectedEdge interface {
```
I don't think that we need to define this. When you have a node, irrespective of whether you have a directed graph or an undirected graph there are outgoing edges from that node. The difference then is that an undirected graph returns all incident edges, wherase a directed graph only returns edges that face out. To handle the semantics of this, undirected graphs should always return an edge{u,v} for a call to Out(u) and always return edge{v,u} for a call to In(u).

This leaves the problem of how you distinguish between directed and undirected graphs - my api-explore branch shows how this can work; define EdgeBetween and EdgeFromTo methods. Now, we have that graph.Graph interface with the In and Out methods that you introduce here.

kortschak commented on graph.go:
```diff
@@ -11,197 +11,162 @@ type Node interface {
-// Returns all undirected edges in the graph
-type EdgeLister interface {
-	EdgeList() []Edge
+type FiniteGraph interface {
```
This is at the centre of something that has not been properly discussed; the goals and non-goals of the graph package.

All our graphs are finite. We have a contract that our nodes are uniquely defined in the context of the graph they are being used in. This limits the number of nodes we can have to max(uint) even if we are not returning a slice of the graph's nodes.

I think that by trying to accommodate true infinite graphs we are significantly weakening the value of the package for more common uses. There are far better languages for dealing with theoretical graphs. Handling infinite graphs (not to say implicitly defined or dynamically generated graphs) should be a non-goal for the package.

Jragonmiris replied:
You're right, infinite graphs shouldn't really be a "thing". I keep using "finite" when I really mean something more like "graph with easily enumerable vertices and edges." All graphs we're handling are finite, but I deal with a ton of dynamically generated graphs where you may not even know an upper bound other than "finite" that I'd like to use the package on.

kortschak replied:
I think my iterator approach deals with those. Are you happy with that?

kortschak commented:
The API looks very much more complicated in this state.

I'm not wildly happy that tests have been removed as a part of the change. IMO, more tests should be added prior to making such widespread changes to help prevent those changes from introduce new bugs.

I would really like to have more discussion about what the use cases are that provoke the changes that we will make - we really need to define the goals and non-goals for the package. Once we have that clear we can make informed decisions about what directions to take. When we do do that I would much rather seen a progressive adoption of smaller interfaces (possible working on one algorithms or family of algorithms) at a time. Go really lends itself towards this kind of change.

The changes that I have been experimenting with on the api-explore branch (not yet in a state that I am yet completely happy with and not ever for submission - though I hope that some of the states there in the final commit will be brought back to the master branch) will get us into a state where we can explore what are the best approaches to take, with minimal API disturbance and with generally small diffs. This has worked for us well in blas, matrix and optimize.

Jragonmiris commented:
I wouldn't say tests have been "removed". They were unfortunately put in limbo because we can't test without concrete graphs and the breaking changes broke all the concrete graphs at an interface level; I decided to commit the interface for RFC rather than fixing the concretes and tests because I had to go to bed when I pushed this. (Same reason I removed Johnson's -- it required a mutable graph and I had no idea how mutability would fit into the experimental structure at that moment).

As for the reasons I made the changes I did:

1. I changed the signatures so much because I thought the cost function stuff was fairly confusing.
2. I made the interfaces so small because when using this I've frequently felt myself doing things like

    func (g *MyGraph) Predecessors(n *graph.Node) []graph.Node {
        panic("Unsupported")
    }

I did probably get too overzealous, but I wanted to at least see what the package might look like if each algorithm were to only require exactly what it needed. I wasn't really thinking of committing what I had without radical changes either, but this branch was a blasting ground where I wanted to see what the package looked like after blowing things up and putting the pieces back together.

I'm curious as to why you think it looks much more complicated. I agree it's difficult to keep track of the interfaces and that needs work, but the algorithms got simpler due to drastically minimizing switching behavior, and each algorithm has a much more minimal set of input constraints and parameters now.

kortschak commented:
For me to be able to comment on the API truly constructively I need to be able to understand it and the motivations behind the changes. I don't think I can do this here because the changes are very big and I don't have a road map for where it is going. What makes it worse for me is that I really need to see that things continue to work (and how the code will be used) - so tests really do need to stay while the changes are made (or at least before I can make comments that are likely to be helpful in moving forward).

I can understand the motivation for reducing the sizes of the consumed interfaces (this was why I answered "Yay" to #41). However, I think that that can be attacked by altering the consumed interfaces rather than the graph package-defined interfaces. This is where I'm heading with api-explore (though at the same time changing names to be more Go-ish). So for example, take the traverse Walk methods; both of these could operate on an `interface { Out(Node) []Edge }` (I'd call this `Source` with its inverse being `Sink`) or near equivalent. If graph.Graph has this method, then the WalkAll would sill be able to pass what it was given to Walk (and Walk could potentially type switch for other methods within Graph for possible optimisations - `interface { Neighbors() []Node; EdgeBetween(u, v Node) Edge }` is more efficient than a pure `Out(Node) []Edge` since it avoids redundant Edge collection. You can see that using this approach, we can progressively restrict the consumed interfaces while retaining the broad graph interface definition. At a later date, when we are comfortable with what really constitutes a minimal Graph interface and settle on that before an API freezish.
